### PR TITLE
SITES-50677: Handle composite Content Fragment fields gracefully

### PR DESCRIPTION
**Pull request by stefangrimm**
| Q                        | A
| ------------------------ | ---
| Fixed Issues?            | Jira: [SITES-50677](https://jira.corp.adobe.com/browse/SITES-50677) (no GitHub issue)
| Patch: Bug Fix?          | Yes — composite (nested/structured) Content Fragment fields produced no usable output anywhere they were touched
| Minor: New Feature?      | No
| Major: Breaking Change?  | See "Versioning" below — a deliberate, risk-based exception to stay in v1; **needs explicit maintainer sign-off**
| Tests Added + Pass?      | Yes — 1181 tests, 0 failures, 0 errors (`mvn clean test` on `bundles/core`), plus a new content-module guardrail test
| Documentation Provided   | No
| Any Dependency Changes?  | No
| License                  | Apache License, Version 2.0

## What this does

Composite (nested/structured) Content Fragment fields are now handled explicitly instead of silently. The Content Fragment and Content Fragment List core components read fragments through the base Content Fragment API and assumed every field was scalar; when a composite field is present (behind an upstream feature toggle in the Content Fragment platform), the component didn't crash, but it didn't handle the field properly either — empty value on the page, an unusable entry in `.model.json`, and a selectable-but-useless entry in edit-dialog pickers.

1. **`.model.json` export** (`DAMContentFragmentImpl`) — the composite element is now omitted entirely from both the exported elements map (`elements`) and the element order (`elementsOrder`). It's still present in `getElements()` (consumed by HTL/data layer) so the render hint below still works.
2. **Page render** (`element.html`) — a composite field shows a localized "this field type is not supported" hint in AEM edit/preview mode, and nothing at all (title, wrapper, value) on publish.
3. **Content Fragment's own element picker** (`ElementsDataSourceServlet`) — composite fields are excluded in both display modes (default and `singleText`); the `singleText` mode had its own separate leak via a `text/plain` content-type fallback.
4. **Content Fragment List's own "Elements"/"Order By" pickers** (`ModelElementsDataSourceServlet`, shared by v1 and v2 dialogs) — same exclusion, via a different mechanism: this servlet has no CFM instance API, so it reads a model's dialog field definitions directly and now checks the field's `valueType` property (the only property here that reliably identifies a field's actual data type — see item 5).
5. Both the Content Fragment and Content Fragment List components share items 1–3 via one underlying model class (`DAMContentFragmentImpl`); item 4 is List-specific dialog code, fixed separately since it doesn't go through that shared class.
6. Analytics data layer output is left unchanged — it has no field-type marker today, so a composite field is indistinguishable from an empty text field there either way; deferred to full composite support.

## Versioning

Stays in **v1** — no new major version. This reverses a literal reading of the versioning policy (the `.model.json` shape and HTL markup both technically change), justified by composite fields being toggle-gated and never having produced usable output in v1. This is a deliberate exception, not an oversight, and **requires maintainer sign-off** before this is merge-ready.

## Testing

- Full `bundles/core` unit test suite: 1181 tests, 0 failures, 0 errors.
- New/updated coverage: `ContentFragmentImplTest`, `ContentFragmentListImplTest`, `ElementsDataSourceServletTest`, `ModelElementsDataSourceServletTest` (new composite-field cases), plus a source-contract guardrail test (`content/test/guardrails/contentFragmentCompositeRendering.js`) locking in the `element.html`/`templates.html` structure, since no Java/IT harness renders that template directly.
- Manual end-to-end verification on a local AEM instance with `FT_COMPOSITE_ELEMENTS` enabled: hint text in edit/preview, nothing on publish, composite absent from both components' pickers, composite absent from `.model.json` — confirmed working as expected.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/servlets/contentfragment/ModelElementsDataSourceServlet.java [hunk 1]
```diff
@@ -123,8 +123,10 @@
                         valueValue = "jcr:content/data/master/" + valueValue;
                     }
                     String metaType = valueMap.get("metaType", StringUtils.EMPTY);
-                    if (!isOrderBy || StringUtils.startsWith(metaType, "text-") || StringUtils.equals(metaType, "date")
-                        || StringUtils.equals(metaType, "number")) {
+                    String valueType = valueMap.get("valueType", StringUtils.EMPTY);
+                    boolean isComposite = "composite".equals(valueType) || "composite[]".equals(valueType);
+                    if (!isComposite && (!isOrderBy || StringUtils.startsWith(metaType, "text-") || StringUtils.equals(metaType, "date")
```
[the remark this answers is not in the document]
Comment (stefangrimm) [reply bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/servlets/contentfragment/ModelElementsDataSourceServlet.java:128]: This is existing code (just added the `!isComposite` condition). Let me know if we should change it or if it is ok to keep it as is.

**Review by YahorC — APPROVED** [review]

**Comment by LSantha** [comment]
@stefangrimm, please consider the following points:

## 1. HTL will miss multi-value composites (`composite[]`) — please fix

Java correctly uses `DataType.getValueType()`, which is `"composite"` regardless of cardinality (CFM strips the `[]` and sets `isMultiValue()`). HTL uses `element.dataType`, which is `getTypeString()` and is `"composite[]"` for a multi-value composite.

If that happens:

- the field still renders on publish (empty title/wrapper)
- the "unsupported" hint does not show
- this is the original silent bug, for the cardinality `ModelElementsDataSourceServlet` already special-cases

`isCompositeElement()` should **not** also match `"composite[]"`. That string is the JCR `valueType` encoding; the API has already normalized it away. Dual-matching it on `getValueType()` would be the wrong fix.

**Preferred fix:** add something like `DAMContentElement.isComposite()` and gate HTL on that. Do not teach HTL the deprecated type-string format.

**Minimum fix:** handle both `'composite'` and `'composite[]'` in every HTL condition, and add a `composite[]` fixture. The new guardrail currently locks in the `== 'composite'` literal, so it needs updating either way.

## 2. `isCompositeElement` can NPE — please null-guard

`element.getValue()` is `@Nullable`. Today composites happen to return a non-null `FragmentData`, but that is not an API contract. A null here takes down the whole CF model / picker.

```java
FragmentData data = element.getValue();
return data != null
    && data.getDataType() != null
    && "composite".equals(data.getDataType().getValueType());
```

Please add unit tests, including `getValue() == null`. `ContentFragmentUtilsTest` does not cover the new method at all.

## 3. Tests only cover `valueType: "composite"` on the API/HTL path

`test-content.json` already has `multiCompositeField` with `"valueType": "composite[]"`, but that only exercises `ModelElementsDataSourceServlet` (raw JCR). The mock currently sets `getValueType()` and `getTypeString()` from the same string, so it cannot catch the mismatch in (1).

Please add a `composite[]` model + DAM fixture through export, picker, and HTL. Split the mock so the two methods can disagree.

## 4. v1 versioning exception — needs explicit maintainer sign-off

Omitting keys from `.model.json` / `elementsOrder` is a breaking change for any consumer that iterated all elements. The justification (toggle-gated, never produced usable output) is coherent, but the PR already flags this as required. Don't merge without that ack.

## Nits (non-blocking)

- CodeQL on existing `StringUtils.startsWith` / `equals` in `ModelElementsDataSourceServlet` — fine to leave.
- `getEditorJSON` still echoing authored composite names is OK (old configs); a one-line comment would help.
- Three different "is composite?" encodings (Java API, HTL type string, model JCR `valueType`) — collapsing via `isComposite()` would help.

**Comment by stefangrimm** [comment]
Thanks for the detailed review! I dug into #1 and don't think the multi-value HTL gap exists — details below. Responses to #2-#4 follow.

## Re #1: `element.dataType` is `"composite"` for both cardinalities, not `"composite[]"`

The claim was that `getTypeString()` (which `element.dataType` in HTL maps to, via `DAMContentFragmentImpl.java:233`) returns `"composite[]"` for multi-value composites, while `getValueType()` normalizes to `"composite"`. I traced the actual `cq-dam-cfm-impl` source and don't think that's what happens — both end up `"composite"` for the real field/element `DataType` object:

- `FieldImpl`'s constructor sets `this.dataType = calculateDataType(scaffoldDef)`, calling `FieldImpl`'s own private, `Resource`-based method (`FieldImpl.java:253-299`).
- That method strips the `[]` suffix into `isMultiValue` (line 262-265), then computes the final `typeStr` via `handleTypeStringBC(semanticType, valueType)` (line 294, method body at line 320+). For composite, `semanticType` is always `null`, so `handleTypeStringBC` falls through to `return valueType;` — meaning **`typeStr` also ends up `"composite"`**, with no `[]`, for both single- and multi-value.
- `CompositeFieldImpl extends FieldImpl` and adds no data-type logic of its own (`CompositeFieldImpl.java:34-38`) — it just calls `super(scaffoldDef, toggleRouter)`, so it inherits exactly this computation.
- `PropertyElement.getDataType()` / `CompositeElementImpl.getDataType()` (the latter extends the former) both just return `field.getDataType()` — i.e. the already-normalized `DataType` from the step above.

There *is* a related inconsistency in `DataTypeUtils.calculateDataType(String)` (a sibling overload) — it preserves the raw `originalTypeStr` (including `[]`) as its `typeString`, without running it through `handleTypeStringBC`. But that overload is only called once in the whole codebase, from `ModelImpl.processScaffold()`, purely to decide whether to instantiate `CompositeFieldImpl` vs. `FieldImpl` (via `getValueType()`, which *is* stripped correctly there) — the resulting `DataType` object is discarded immediately after that check and never attached to the field or element. So it's latent tech debt in `DataTypeUtils`, but it never reaches `element.dataType` in HTL.

Given that, I'd rather not add the `'composite[]'` string match — per the trace above, `getTypeString()` never actually produces that value for a real element, so matching it would just be dead code (and would risk implying it *could* legitimately mean something else if the platform's internals ever change to emit it).

Let me know if you disagree or have a repro where `element.dataType` actually comes back as `"composite[]"` on a real fragment — that would mean I'm missing a code path.

## Re #2: NPE guard on `getValue()`

Checked `ContentElement.getValue()` in the CFM API sources — it's **not** annotated `@Nullable` (other methods in the same interface, e.g. `DataType.getSemanticType()`, are explicitly marked `@Nullable` where that's the actual contract; `getValue()` isn't one of them). I also checked: no existing call site in this codebase null-guards `getValue()` either (e.g. `DAMContentFragmentImpl.java:257` calls `value.getDataType().getTypeString()` on the same kind of value with no check).

Given that, I don't think this is a real gap in the API contract, and adding a guard here — while leaving every other `getValue()` call site unguarded — would be an inconsistent, one-off pattern rather than a fix. I'd rather not add it. If you know of a concrete path where `getValue()` returns `null` for a composite element (vs. it just being theoretically possible from the method signature alone), let me know and I'll reconsider.

## Re #3: test coverage for `valueType: "composite[]"` through the API/HTL path

Since #1 doesn't hold, there's no existing bug for this to catch. But I take the underlying point: the current mock setup couples `getValueType()` and `getTypeString()` to the same source string, so a future regression where they *do* diverge (e.g. if someone "fixes" `DataTypeUtils.calculateDataType(String)`'s inconsistency in a way that leaks into `FieldImpl` too) would go undetected today.

Rather than adding the fixture the way originally suggested (which would imply the bug exists today), I'll add a regression test that explicitly asserts `getTypeString()` and `getValueType()` both normalize to `"composite"` for a `composite[]`-valued element through `ContentFragmentUtils.isCompositeElement()` and the HTL-facing `getDataType()` path — locking in the invariant this whole discussion turned on, rather than guarding a hypothetical bug. Pushing that as a follow-up commit now.

## Re #4: versioning sign-off

Agreed — already called out in the PR description as blocking; not merging without explicit maintainer sign-off.

**Comment by vladbailescu** [comment]
@stefangrimm , please consider adding an example CF with composite field in the examples/library and adding a CF component that uses it on the corresponding library page! It will be both an example to showcase expected behavior and also good for manual validation.

File: bundles/core/src/test/java/com/adobe/cq/wcm/core/components/internal/ContentFragmentUtilsTest.java [hunk 2]
```diff
@@ -327,6 +329,36 @@
...
+
+    /**
+     * Mocks a {@link ContentElement} whose {@link FragmentData#getDataType()} independently
+     * controls {@code valueType} and the deprecated {@code typeString}, so tests can prove which
+     * one production code actually depends on.
+     */
+    private static ContentElement mockElementWithDataType(String valueType, String typeString) {
+        DataType dataType = Mockito.mock(DataType.class);
+        Mockito.when(dataType.getValueType()).thenReturn(valueType);
+        Mockito.when(dataType.getTypeString()).thenReturn(typeString);
```
[the remark this answers is not in the document]
Comment (stefangrimm) [reply bundles/core/src/test/java/com/adobe/cq/wcm/core/components/internal/ContentFragmentUtilsTest.java:352]: This is a test for existing code that uses the deprecated method -> required.

**Comment by stefangrimm** [comment]
@vladbailescu This PR is about preventing the component to silently crash for fragments with composite data type, so the demo content would show nothing special. (Testing might be a valid reason for adding the content, though.) Furthermore, the composite data type is not (yet) implemented for 6.5 (but AFAIU Core Components are supposed to support it), so there's a certain risk that something breaks when demo data with composite fields are added.

Given this background - do you still want me to add that demo content?

**Comment by stefangrimm** [comment]
@LSantha FYI - I triple-checked (in code and on a real instance) that the feedback is false-positive. See the previous comments for a full explanation.

**Comment by LSantha** [comment]
@stefangrimm , 1. and 2. were indeed false positives. So the only remaining question to @vladbailescu and to you is wether there are real chances that this would break existing customers relying on model.json structure. or markup? I give my approval for the implementation, the question of v2 component is open.

**Review by LSantha — APPROVED** [review]

**Comment by stefangrimm** [comment]
@LSantha The composite data type is not GA yet, so currently there are no instances with fragments that contain  composite data (besides a small group of customers we fully control - and if those use composite and core components, they opt out, as it is not supported properly (and will not, even with this fix)). There is no need for BC, as there are no occurrences of an "old state" around in the wild. For AEM, I would say this is safe.

However, AFAIU Core Components have a release cycle that is independent of AEM. So let's assume composite data gets GAed around the end of the year, and a customer uses them, but does not update to the latest version of core components (a version that doesn't contain this PR).

In this case, that customer will:

- see composite fields in model.json's `elementOrder`
- the field will be exposed as a `null` value (also model.json)
- in HTL, composite fields will appear (element name as headline, content is "empty")
- in the dialogs, composite fields appear in element selectors and can be selected

Now, if that customer updates to the latest version of core components (including this PR):

- composite fields will not appear at all in model.json
- also, in HTL the composite elements will show as "not supported" on author and be omitted on "publish"
- composite fields will not appear in element selectors in dialogs

From my perspective, this is too much of an edge case to warrant a new v2. I tend to handle it more using release notes ("if you use composite fields, ensure you're using the latest version of Core Components") and documentation, but I am also aware customers tend to not read this.

The chance code processing model.json actually breaks is small, IMHO - the much bigger problem is that you get a `null` instead of the real value, even with this fix. The HTL rendering difference is real, though - however, the "old" rendering doesn't make much sense either.

But in theory there is a very small chance something breaks IMHO. Usually we would handle this with feature toggles, but AFAIU this is not available here.

**So it's up to you to decide** if the given context requires a v2.

If we need a v2, we should however evaluate if it is worth it for this "graceful degradation" fix, or if we should simply go with the current v1, which is not handling composites nicely, but it also doesn't break. A v2 should provide full support for composites IMHO.

**Comment by vladbailescu** [comment]
@stefangrimm , I'm ok to go with the current approach and not create a new version as v1 users might still experience breaking behavior if they use it with a composite data fragment. Also, I understand your concern related to adding an example that might not working on 6.5. I had the impression that composite field content would be just ignored on 6.5 (if for example it is installed there from a package created on Cloud Service); I'm fine to skip adding the example.

**Comment by stefangrimm** [comment]
@vladbailescu I did a smoke test with a 6.5 instance and fragments with composite fields. It failed when I installed a CS content package completely, due to a missing nodetype. This is a general problem - you currently can't install fragments (with or without composite fields) originating from CS on a 6.5 (LTS) instance due to that missing nodetype.

However, I think the nodetype is for some advanced feature (semantic search IIRC), so I could remove it from my demo content package. This kind of example content can't be used for the respective feature (semantic search in this case), but if I check out the existing example fragments (like `examples/ui.content/src/content/jcr_root/content/dam/core-components-examples/library/sample-assets/adobe-headquarters/.content.xml`), they are on a rather old content version anyway (e.g. no support for efficient GraphQL filtering), so it looks like this trade-off is accepted and it is ok if not all features are supported.

Once I made the fragments compatible, they installed and at least superficially worked. However, if you open and save a model, it will remove the composite fields. The fragment editor seems to handle it more gracefully - the worst thing I saw was that the composite fields appeared in the structure view - more a cosmetical issues.

There might be other hidden problems (I only smoke tested it), and those exist in CS as well - we never designed the feature in a way that it is used like that. Even on instances where the toggle was on, content was created, and then switched off again, a manual cleanup is required, although CS has a more mechanisms built in that protect the system.

TL;DR - After doing some research I stand by my suggestion that the risks of introducing demo content with composite fields outweigh the benefits, at least for the current implementation.
